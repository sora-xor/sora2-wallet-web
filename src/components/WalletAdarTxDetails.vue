<template>
  <div v-if="swapTransferBatchRecipients.length > 0" class="adar-tx-details">
    <div class="adar-tx-details__title">
      {{ t('transaction.adarTxDetailsTitle') }}
    </div>
    <div class="adar-tx-details__txs-container">
      <div v-for="(recipient, idx) in txsList" :key="idx">
        <transaction-hash-view
          :translation="'transaction.to'"
          :value="recipient.accountId"
          :type="HashType.Account"
          class="adar-tx-details__account-id"
        />
        <info-line
          is-formatted
          value-can-be-hidden
          :label="t('transaction.amount')"
          :value="recipient.amount"
          :asset-symbol="recipient.symbol"
        />
      </div>
    </div>
    <s-pagination
      class="adar-tx-details__pagination-panel"
      :layout="'prev, total, next'"
      :current-page.sync="currentPage"
      :page-size="pageAmount"
      :total="swapTransferBatchRecipients.length"
      @prev-click="handlePrevClick"
      @next-click="handleNextClick"
    />
  </div>
</template>

<script lang="ts">
import { Operation } from '@sora-substrate/util';
import { Component, Mixins, Prop } from 'vue-property-decorator';

import { SwapTransferBatchTransferParam } from '@/services/indexer/types';
import { formatAddress } from '@/util';

import { HashType } from '../consts';
import { getter } from '../store/decorators';

import InfoLine from './InfoLine.vue';
import NumberFormatterMixin from './mixins/NumberFormatterMixin';
import PaginationSearchMixin from './mixins/PaginationSearchMixin';
import TranslationMixin from './mixins/TranslationMixin';
import TransactionHashView from './TransactionHashView.vue';

import type { PolkadotJsAccount } from '../types/common';
import type { HistoryItem } from '@sora-substrate/util';

@Component({
  components: {
    InfoLine,
    TransactionHashView,
  },
})
export default class WalletAdarTxDetails extends Mixins(TranslationMixin, NumberFormatterMixin, PaginationSearchMixin) {
  readonly HashType = HashType;

  @getter.account.account private account!: PolkadotJsAccount;

  @Prop() readonly transaction!: HistoryItem;

  pageAmount = 4;

  get isAdarOperation(): boolean {
    return this.transaction.type === Operation.SwapTransferBatch;
  }

  get swapTransferBatchAmount(): string {
    if (!this.isAdarOperation) return '0';
    const isRecipient = this.account.address !== this.transaction.from;
    if (isRecipient) {
      const amount = this.transaction.payload.transfers.find(
        (transfer: SwapTransferBatchTransferParam) => transfer.to === this.account.address
      )?.amount;
      return this.formatStringValue(amount);
    } else {
      return this.formatStringValue(this.transaction.amount || '0');
    }
  }

  get swapTransferBatchRecipients() {
    if (!this.isAdarOperation || this.account.address !== this.transaction.from) return [];
    return this.transaction.payload.receivers;
  }

  get txsList() {
    return this.getPageItems(this.swapTransferBatchRecipients);
  }

  formatAddress(address: string) {
    return formatAddress(address);
  }
}
</script>

<style lang="scss">
.adar-tx-details {
  &__pagination-panel {
    display: flex;
    margin-top: $inner-spacing-medium;
    .el-pagination__total {
      margin: auto;
    }
  }
}
</style>

<style scoped lang="scss">
.adar-tx-details {
  &__txs-container {
    & > * {
      margin-bottom: #{$basic-spacing-big};
    }
  }

  &__account-id {
    margin-bottom: #{$basic-spacing-small};
  }

  &__title {
    font-size: var(--s-font-size-big);
    text-align: center;
    margin: #{$basic-spacing-medium} 0;
    text-transform: capitalize;
  }
}
</style>
