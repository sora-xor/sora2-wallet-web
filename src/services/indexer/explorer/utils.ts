import { formatStringNumber } from '../../../util';

import type { SubqueryAssetEntity } from '../subquery/types';
import type { SubsquidAssetEntity } from '../subsquid/types';
import type { FiatPriceObject, UpdatesStream, PoolApyObject } from '../types';

export function parseAssetFiatPrice(entity: SubsquidAssetEntity | SubqueryAssetEntity): FiatPriceObject {
  const acc = {};
  const id = entity.id;
  const priceFPNumber = formatStringNumber(entity.priceUSD);
  const isPriceFinity = priceFPNumber.isFinity();
  if (isPriceFinity) {
    acc[id] = priceFPNumber.toCodecString();
  }
  return acc;
}

export function parsePriceStreamUpdate(entity: UpdatesStream): Nullable<FiatPriceObject> {
  if (!entity?.data) return null;

  const data = JSON.parse(entity.data);

  return Object.entries(data).reduce((acc, [id, price]) => {
    const priceFPNumber = formatStringNumber(price as string);
    const isPriceFinity = priceFPNumber.isFinity();
    if (isPriceFinity) {
      acc[id] = priceFPNumber.toCodecString();
    }
    return acc;
  }, {});
}

export function parseApyStreamUpdate(entity: UpdatesStream): PoolApyObject {
  const data = entity?.data ? JSON.parse(entity.data) : {};

  return Object.entries(data).reduce((acc, [id, apy]) => {
    const strategicBonusApyFPNumber = formatStringNumber(apy as string);
    const isStrategicBonusApyFinity = strategicBonusApyFPNumber.isFinity();
    if (isStrategicBonusApyFinity) {
      acc[id] = strategicBonusApyFPNumber.toCodecString();
    }
    return acc;
  }, {});
}
