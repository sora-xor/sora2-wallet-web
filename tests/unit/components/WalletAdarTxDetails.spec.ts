import { Operation, HistoryItem } from '@sora-substrate/sdk';

import WalletAdarTxDetails from '@/components/WalletAdarTxDetails.vue';

import { useDescribe, useShallowMount, useVuex } from '../../utils';
import { MOCK_ACCOUNTS, MOCK_ACCOUNT_ASSETS, MOCK_HISTORY } from '../../utils/mock';

const createStore = (tx: HistoryItem) =>
  useVuex({
    router: {
      state: () => ({
        currentRouteParams: { id: '1', asset: MOCK_ACCOUNT_ASSETS[0] },
      }),
      mutations: {
        navigate: jest.fn(),
      },
    },
    account: {
      state: () => ({
        accountAssets: MOCK_ACCOUNT_ASSETS,
      }),
      getters: {
        account: () => MOCK_ACCOUNTS[0],
      },
    },
  });

useDescribe('WalletAdarTxDetails.vue', WalletAdarTxDetails, () => {
  MOCK_HISTORY.filter((item) => item.type === Operation.SwapTransferBatch).map((item) => [
    it(`[${item.type}, ${item.status}]: should be rendered correctly`, () => {
      const wrapper = useShallowMount(WalletAdarTxDetails, {
        store: createStore(item),
        propsData: { transaction: item },
      });
      expect(wrapper.element).toMatchSnapshot();
    }),
  ]);
});
