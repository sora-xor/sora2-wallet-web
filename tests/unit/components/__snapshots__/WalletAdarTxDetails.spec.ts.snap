// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletAdarTxDetails.vue [SwapTransferBatch, finalized]: should be rendered correctly 1`] = `
<div
  class="adar-tx-details"
>
  <div
    class="adar-tx-details__title"
  >
    
    Transaction details
  
  </div>
   
  <div
    class="adar-tx-details__txs-container"
  >
    <div>
      <transaction-hash-view-stub
        block=""
        class="adar-tx-details__account-id"
        hash=""
        translation="transaction.to"
        type="account"
        value="cnSmwADqUFSbXA5iGZYizXwCwNPHHt7dTG2ZApDQPkq92TFCT"
      />
       
      <info-line-stub
        assetsymbol="XOR"
        fiatvalue=""
        isformatted="true"
        label="Amount"
        labeltooltip=""
        value="0.732178"
        valuecanbehidden="true"
        valuetooltip=""
      />
    </div>
    <div>
      <transaction-hash-view-stub
        block=""
        class="adar-tx-details__account-id"
        hash=""
        translation="transaction.to"
        type="account"
        value="cnV2xJFCvXN9fx3DHt9FJWovPkVVBu1vtuzmhr2QnedmZqQ6J"
      />
       
      <info-line-stub
        assetsymbol="VAL"
        fiatvalue=""
        isformatted="true"
        label="Amount"
        labeltooltip=""
        value="0.8836505"
        valuecanbehidden="true"
        valuetooltip=""
      />
    </div>
  </div>
   
  <s-pagination-stub
    borderradius="small"
    class="adar-tx-details__pagination-panel"
    currentpage="1"
    layout="prev, total, next"
    nexttext=""
    oftext="of"
    pagesize="4"
    pagesizes="10,20,30,40,50,100"
    perpagetext="Rows per page"
    popperclass=""
    prevtext=""
    total="2"
    totaltext="Total"
  />
</div>
`;
